src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs

```diff
@@ -0,0 +1,589 @@
+        {
+            if (propertyName.Length > JsonConstants.MaxCharacterTokenSize || value.Length > JsonConstants.MaxCharacterTokenSize)
+                ThrowHelper.ThrowArgumentException(propertyName, value);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int CountDigits(ulong value)
+        {
+            int digits = 1;
+            uint part;
+            if (value >= 10000000)
+            {
+                if (value >= 100000000000000)
+                {
+                    part = (uint)(value / 100000000000000);
+                    digits += 14;
+                }
+                else
+                {
+                    part = (uint)(value / 10000000);
+                    digits += 7;
+                }
+            }
+            else
+            {
+                part = (uint)value;
+            }
+
+            if (part < 10)
+            {
+                // no-op
+            }
+            else if (part < 100)
+            {
+                digits += 1;
+            }
+            else if (part < 1000)
+            {
+                digits += 2;
+            }
+            else if (part < 10000)
+            {
+                digits += 3;
+            }
+            else if (part < 100000)
+            {
+                digits += 4;
+            }
+            else if (part < 1000000)
+            {
+                digits += 5;
+            }
+            else
+            {
+                Debug.Assert(part < 10000000);
+                digits += 6;
+            }
+
+            return digits;
+        }
+
+        /// <summary>
+        /// Converts a span containing a sequence of UTF-16 bytes into UTF-8 bytes.
+        ///
+        /// This method will consume as many of the input bytes as possible.
+        ///
+        /// On successful exit, the entire input was consumed and encoded successfully. In this case, <paramref name="bytesConsumed"/> will be
+        /// equal to the length of the <paramref name="source"/> and <paramref name="bytesWritten"/> will equal the total number of bytes written to
+        /// the <paramref name="destination"/>.
+        /// </summary>
+        /// <param name="source">A span containing a sequence of UTF-16 bytes.</param>
+        /// <param name="destination">A span to write the UTF-8 bytes into.</param>
+        /// <param name="bytesConsumed">On exit, contains the number of bytes that were consumed from the <paramref name="source"/>.</param>
+        /// <param name="bytesWritten">On exit, contains the number of bytes written to <paramref name="destination"/></param>
+        /// <returns>A <see cref="OperationStatus"/> value representing the state of the conversion.</returns>
+        public unsafe static OperationStatus ToUtf8(ReadOnlySpan<byte> source, Span<byte> destination, out int bytesConsumed, out int bytesWritten)
```

Comment:
Is this method really necessary compared to the normal Encoder methods? If yes, maybe move it to an extra file? This is a very specific implementation, with important comments and should not be changed individually without good cause, putting it into a different file might help to make sure this is never touched.

Reply:
Nit: `source` => `utf16source`, `destination` => `utf8destination`

Reply:
But I also agree with @Tornhoof: it'd be really good if this could use the existing UTF8 encoder in some way.

Reply:
> This should really be done as code reuse (API or shared source), not code copying.

I agree but this is a point-in-time concern.
I brought in this API (along with the supporting helpers) from corefxlab (https://github.com/dotnet/corefxlab/blob/0abb70efee949cff78a6b485f022dcd2e7896644/src/System.Text.Primitives/System/Text/Encoders/Utf16.cs#L108) to remove the dependency on https://github.com/dotnet/corefx/issues/34094. Ideally, once the Utf8 class is made public, we would replace this implementation (along with having to convert char -> byte while keeping it as UTF-16). I need the operation status based transcoding.
https://github.com/dotnet/corefxlab/blob/0abb70efee949cff78a6b485f022dcd2e7896644/src/System.Text.Primitives/System/Text/Encoders/Utf16.cs#L108

I will add a comment to replace this in the (near?) future.

Reply:
> I need the operation status based transcoding.

Per https://github.com/dotnet/corefx/pull/34425#discussion_r246074152, do we really?

Reply:
> do we really?

I need to be able to do partial writes to the destination buffer since it is possible that the destination is too small for the given input.

I can't do something like the following since it would throw if `_buffer` was too small.
```C#
int written = Encoding.UTF8.GetBytes(escapedValue, _buffer.Slice(idx));
```
I would have to use the pointer overload and calculate how many characters will fit into the leftover destination buffer, up front.

Reply:
these implementations have diverged with all the intrinsic work put into corelib version and comments are outdated https://github.com/dotnet/runtime/blob/eeef4b14c5a137c2f665a9be4f30e4383d1038da/src/libraries/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Transcoding.cs#L28

@eiriktsarpalis can it be replaced now or is it permanent?
btw OperationStatus can be replaced with value tuple throughout this project ([even for .net framework](https://github.com/dotnet/runtime/blob/eeef4b14c5a137c2f665a9be4f30e4383d1038da/src/libraries/System.Text.Json/src/System.Text.Json.csproj#L381)).

Reply:
Not sure, could you create a new issue please?

Reply:
https://github.com/dotnet/runtime/issues/75779